crypto/src/conversation.rs

```diff
@@ -755,6 +747,127 @@ pub mod tests {
+            let (alice_backend, mut alice) = alice();
+            let (charlie_backend, charlie) = charlie();
+            let bob_key = bob.local_client().keypackages(&bob_backend).unwrap()[0].clone();
+            let charlie_key = charlie.local_client().keypackages(&charlie_backend).unwrap()[0].clone();
+
+            let configuration = MlsConversationConfiguration::default();
+
+            let mut alice_group =
+                MlsConversation::create(conversation_id, alice.local_client_mut(), configuration, &alice_backend)
+                    .unwrap();
+
+            let add_message = alice_group.add_members(&mut [bob], &alice_backend).unwrap();
+
+            assert_eq!(alice_group.members().unwrap().len(), 2);
+
+            let MlsConversationCreationMessage { welcome, .. } = add_message;
+
+            let conversation_config = MlsConversationConfiguration::default();
+
+            let mut bob_group =
+                MlsConversation::from_welcome_message(welcome, conversation_config.clone(), &bob_backend).unwrap();
+
+            let bob_keys = bob_group
+                .group
+                .members()
+                .into_iter()
+                .cloned()
+                .collect::<Vec<KeyPackage>>();
+
+            let alice_keys = alice_group
+                .group
+                .members()
+                .into_iter()
+                .cloned()
+                .collect::<Vec<KeyPackage>>();
+
+            assert!(alice_keys
+                .iter()
+                .all(|a_key| bob_keys.iter().any(|b_key| b_key == a_key)));
+
+            let alice_key = alice_keys.into_iter().find(|k| *k != bob_key).unwrap();
+
+            let proposal_response = alice_group
+                .group
+                .propose_add_member(&alice_backend, &charlie_key)
+                .unwrap();
+
+            assert!(bob_group
+                .decrypt_message(&proposal_response.to_bytes().unwrap(), &bob_backend)
+                .unwrap()
+                .is_none());
+
+            assert_eq!(alice_group.group.members().len(), 2);
+
+            let (message, welcome) = alice_group.update_keying_material(&alice_backend).unwrap();
+            assert!(welcome.is_some());
+
+            alice_group.group.merge_pending_commit().unwrap();
+
+            let charlie_welcome = welcome.unwrap();
+            let mut charlie_group =
+                MlsConversation::from_welcome_message(charlie_welcome, conversation_config, &charlie_backend).unwrap();
+
+            assert_eq!(alice_group.members().unwrap().len(), 3);
+            assert_eq!(charlie_group.members().unwrap().len(), 3);
+            assert_eq!(bob_group.members().unwrap().len(), 2);
+
+            let alice_new_keys = alice_group
+                .group
+                .members()
+                .into_iter()
+                .cloned()
+                .collect::<Vec<KeyPackage>>();
+
+            assert!(!alice_new_keys.contains(&alice_key));
```

Comment:
My idea was making sure the key actually changed.

Reply:
Oh my bad I missed the `!` in the test. Anyway, since it's only a group of 2, a self_update of alice should change all the key packages so you should just compare alice's members before & after

Reply:
Actually Bob's keys are not being changed... Printing `alice_keys` and `alice_new_keys` from a test run:

```
---- proposal::proposal_tests::propose_self_update::should_propose_self_update_conversation_group stdout ----
HashReference: 56B62BD6119ABC9C46FE7968BB97BEDF
HashReference: E5976B37EB9358E218EA4DC2042EF618

HashReference: E2F08EDE44D9AE0E626753F82BF08E15
HashReference: E5976B37EB9358E218EA4DC2042EF618
```